### PR TITLE
Swift: Query for bad HTML filtering regexps

### PR DESCRIPTION
**Pull request by geoffw0**
Adds a query for bad HTML filtering regular expressions (built atop the regular expressions library from https://github.com/github/codeql/pull/13470).

~This currently suffers from a high rate of errors in the test, in particular false positive results, I believe because we are not identifying regular expressions with 'mode flags' such as case insensitivity yet.  This will need to be fixed before this PR can be merged.  Nevertheless I want to share my work so far and would welcome any early review comments.~

Uses shared code (from `codeql.regex`) so there isn't a lot of QL to see here.  Reviews will probably focus on results, docs and metadata.  Enhancements have been made to the Swift regex library (outside of this PR) to provide accurate information so that this query gives accurate results.

TODO:
- [x] fix false positive results
  - [x] (and check with MRVA)
- [x] team review
- [x] doc review
- [x] DCA run

### CODE REVIEW

**Comment by geoffw0** [comment]
I've now addressed all the false positives.

All of the test cases beginning with a parse mode flag (such as `(?i)`) are not detected by the query.  As far as I can tell this issue affects all languages, I'm looking into it but I don't think it should hold back the query for Swift.

**Comment by geoffw0** [comment]
DCA run LGTM.  No problems found with MRVA.

File: swift/ql/src/queries/Security/CWE-116/BadTagFilterBad.swift [hunk 1]
```diff
@@ -0,0 +1,9 @@
+let script_tag_regex = /<script[^>]*>.*<\/script>/
+
+var old_html = ""
+while (html != old_html) {
+  old_html = html
+  html.replace(script_tag_regex, with: "")
+}
+
+...
```
Comment (MathiasVP) [inline swift/ql/src/queries/Security/CWE-116/BadTagFilterBad.swift:9]: Shouldn't this be in a `.html` file instead?

**Review by MathiasVP — COMMENTED** [review]
One small comment, otherwise this LGTM once we have a docs review

**Review by MathiasVP — DISMISSED** [review]

**Comment by mchammer01** [comment]
I'll review this for Docs!

**Review by mchammer01 — DISMISSED** [review]
@geoffw0 - this LGTM 🚀 
A few minor comments and suggestions for your consideration (feel free to ignore if you don't agree 😄)

File: swift/ql/src/queries/Security/CWE-116/BadTagFilter.qhelp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+<p>
+The above sanitizer does not filter out all <code>&lt;script&gt;</code> tags.
+Browsers will not only accept <code>&lt;/script&gt;</code> as script end tags, but also tags such as <code>&lt;/script foo="bar"&gt;</code> even though it is a parser error.
+This means that an attack string such as <code>&lt;script&gt;alert(1)&lt;/script foo="bar"&gt;</code> will not be filtered by
+the function, and <code>alert(1)</code> will be executed by a browser if the string is rendered as HTML.
+</p>
+
+<p>
+Other corner cases include that HTML comments can end with <code>--!&gt;</code>,
+and that HTML tag names can contain upper case characters.
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply swift/ql/src/queries/Security/CWE-116/BadTagFilter.qhelp:42]: (you can do a multi-line suggestion in future by drag-selecting both lines before you add a suggestion)

File: swift/ql/src/queries/Security/CWE-116/BadTagFilter.qhelp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+</p>
+</example>
+
+<references>
+<li>Securitum: <a href="https://research.securitum.com/the-curious-case-of-copy-paste/">The Curious Case of Copy &amp; Paste</a>.</li>
+<li>stackoverflow.com: <a href="https://stackoverflow.com/questions/1732348/regex-match-open-tags-except-xhtml-self-contained-tags#answer-1732454">You can't parse [X]HTML with regex</a>.</li>
+<li>HTML Standard: <a href="https://html.spec.whatwg.org/multipage/parsing.html#comment-end-bang-state">Comment end bang state</a>.</li>
+<li>stackoverflow.com: <a href="https://stackoverflow.com/questions/25559999/why-arent-browsers-strict-about-html">Why aren't browsers strict about HTML?</a>.</li>
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply swift/ql/src/queries/Security/CWE-116/BadTagFilter.qhelp:50]: Definitely worth correcting, and very easy to do with your suggested change!

File: swift/ql/src/queries/Security/CWE-116/BadTagFilter.qhelp [hunk 4]
```diff
@@ -0,0 +1,52 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+
+<overview>
+<p>
+It is possible to match some single HTML tags using regular expressions (parsing general HTML using
+regular expressions is impossible). However, if the regular expression is not written well it might
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply swift/ql/src/queries/Security/CWE-116/BadTagFilter.qhelp]: Done.  But let me know if you have a better wording.

[the remark this answers is not in the document]
Comment (mchammer01) [reply swift/ql/src/queries/Security/CWE-116/BadTagFilter.qhelp]: I like your suggestion 👍🏻

**Comment by geoffw0** [comment]
Thanks for the reviews!

Merging.

**Pull request merged**